homeworks/[Homework#8]-JavaScript_Part3_by_Ruslan_Vasylyshyn/solution.js

```diff
@@ -0,0 +1,231 @@
+// 1. Pluck
+function pluck(obj, path) {
+  const keys = path.split(".");
+  let value = obj;
+
+  for (const key of keys) {
+    if (value && typeof value === "object" && key in value) {
+      value = value[key];
+    } else {
+      return null;
+    }
+  }
+
+  return value;
+}
+
+const user = {
+  username: "testuser1",
+  preferences: {
+    sound: {
+      maxValue: 50,
+      value: 30,
+    },
+  },
+};
+const randomValue = Math.random();
+const nullValue = null;
+
+// console.log(pluck(user, 'preferences.sound.value')); // 30
+// console.log(pluck(user, 'unknown.key')); // null
+// console.log(pluck(randomValue, 'unknown.key')); // null
+// console.log(pluck(nullValue, 'unknown.key')); // null
+
+// 2. Deep Clone
+function clone(obj) {
+  if (typeof obj !== "object" || obj === null) {
+    return obj;
+  }
+
+  const clonedObj = {};
+
+  for (let key in obj) {
+    clonedObj[key] = clone(obj[key]);
+  }
+
+  return clonedObj;
+}
+
+const user1 = {
+  username: "testuser1",
+  preferences: {
+    sound: {
+      maxValue: 50,
+      value: 30,
+    },
+  },
+};
+const clonedUser = clone(user1);
+
+clonedUser.preferences.sound.maxValue = 70;
+
+// console.log(
+//   user1.preferences.sound.maxValue === clonedUser.preferences.sound.maxValue
+// ); // false
+
+// 3. "A long time ago"
+// Handled task without third-party library Moment, as Alexander wrote in chat
+// Also added two tests to check minutes and hours
+
+function offset(dateString) {
+  const now = new Date();
+  const inputDate = new Date(
+    dateString.replace(
+      /(\d{2})\/(\d{2})\/(\d{4}) (\d{2}):(\d{2}):(\d{2})/,
+      "$3-$2-$1T$4:$5:$6"
+    )
+  );
+  const diffMilliseconds = now - inputDate;
+  const diffSeconds = Math.floor(diffMilliseconds / 1000);
+  const diffMinutes = Math.floor(diffSeconds / 60);
+  const diffHours = Math.floor(diffMinutes / 60);
+  const diffDays = Math.floor(diffHours / 24);
+
+  if (isNaN(diffSeconds)) {
+    return "Invalid date";
+  }
+
+  if (diffDays > 0) {
+    return `${diffDays} day${diffDays > 1 ? "s" : ""} ago`;
+  } else if (diffHours > 0) {
+    return `${diffHours} hour${diffHours > 1 ? "s" : ""} ago`;
+  } else if (diffMinutes > 0) {
+    return `${diffMinutes} minute${diffMinutes > 1 ? "s" : ""} ago`;
+  } else {
+    return `${diffSeconds} second${diffSeconds > 1 ? "s" : ""} ago`;
+  }
+}
```

Comment:
Didn't got the main result, cause you get as a params date string and offset and need to calculate using both params

Reply:
> Didn't got the main result, cause you get as a params date string and offset and need to calculate using both params

As far as I understand, the second parameter is used by the Moment library to format the date. Since I wasn't using this library, there was no need for additional formatting, so I didn't add this option. 
Am I understanding you correctly, I need to add the ability to format the date with the second parameter?

Reply:
I see
No, you can leave as it is

Reply:
And you can merge

Reply:
Ok, thanks